File: src/main/java/net/alexwells/roomery/tile/TileEntityBase.java
package net.alexwells.roomery.tile;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.play.server.SPacketUpdateTileEntity;
import net.minecraft.tileentity.TileEntity;

public class TileEntityBase extends TileEntity {
    protected void sendUpdatePacket() {
        if(world.isRemote) {
            return;
        }

        SPacketUpdateTileEntity updatePacket = getUpdatePacket();

        if(updatePacket == null) {
            return;
        }

        for(EntityPlayer player : world.playerEntities) {
            if (Math.hypot(player.posX - pos.getX() + 0.5, player.posZ - pos.getZ() + 0.5) < 64) {
                ((EntityPlayerMP) player).connection.sendPacket(updatePacket);
            }
        }
    }

    protected boolean useDefaultUpdatePacket() {
        return false;
    }

    @Override
    public SPacketUpdateTileEntity getUpdatePacket() {
        if(!useDefaultUpdatePacket()) {
            return null;
        }

        NBTTagCompound tagCompound = new NBTTagCompound();
        writeToNBT(tagCompound);

        return new SPacketUpdateTileEntity(pos, 1, tagCompound);
    }

    @Override
    public void onDataPacket(NetworkManager net, SPacketUpdateTileEntity pkt) {
        readFromNBT(pkt.getNbtCompound());

        world.markBlockRangeForRenderUpdate(pos, pos);
    }
}

File: src/main/java/net/alexwells/roomery/tile/TileEntityBase.kt
package net.alexwells.roomery.tile

import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.NetworkManager
import net.minecraft.network.play.server.SPacketUpdateTileEntity
import net.minecraft.tileentity.TileEntity
import net.minecraft.tileentity.TileEntityType

abstract class TileEntityBase(tileEntityTypeIn: TileEntityType<*>) : TileEntity(tileEntityTypeIn) {
    protected fun sendUpdatePacket() {
        if (world.isRemote) {
            return
        }

        val updatePacket = updatePacket ?: return

        for (player in world.playerEntities) {
            if (Math.hypot(player.posX - pos.x + 0.5, player.posZ - pos.z + 0.5) < 64) {
                (player as EntityPlayerMP).connection.sendPacket(updatePacket)
            }
        }
    }

    protected open fun useDefaultUpdatePacket() = false

    override fun getUpdatePacket(): SPacketUpdateTileEntity? {
        if (!useDefaultUpdatePacket()) {
            return null
        }

        val tagCompound = NBTTagCompound()
        write(tagCompound)

        return SPacketUpdateTileEntity(pos, 1, tagCompound)
    }

    override fun onDataPacket(net: NetworkManager, pkt: SPacketUpdateTileEntity) {
        read(pkt.nbtCompound)
    }
}
